### PR TITLE
Diffie hellman parsing

### PR DESCRIPTION
**Pull request by jcm-cryptosense**

### CODE REVIEW

**Comment by NathanReb** [comment]
Can you please rebase on top of master ? I fixed the dependencies issue.

File: src/asn1.ml [hunk 1]
```diff
@@ -528,6 +596,17 @@ struct
     map f g @@ sequence2
       (required ~label:"algorithm" Algo.grammar)
       (required ~label:"parameters" EC.Params.grammar)
+
+  let dh_grammar =
+    let open Asn in
+    let f = function
+      | Algo.DH, params -> params
+      | Algo.DH, _ -> parse_error "Algorithm OID valid, but something else went wrong."
```
Comment (NathanReb) [inline src/asn1.ml]: I'd say drop this line since this error message won't make sense when using the generic `decode`.

**Comment by NathanReb** [comment]
Could you please add a dh case to `X509.decode`, I think you forgot about it !

**Comment by NathanReb** [comment]
Looks good, just add some tests and we'll merge !

File: tests/test_asn1.ml [hunk 2]
```diff
@@ -273,8 +273,80 @@ let ec_suite =
   ; "PKCS8" >::: pkcs8_suite
   ]
 
+let dh_suite =
+  let open Key_parsers.Asn1.DH in
+  (* These parameters and key pair were generated using openssl dsaparam and gendsa *)
```
Comment (NathanReb) [inline tests/test_asn1.ml]: I think you mean `dhparam` and `genpkey`

Comment (jcm-cryptosense) [inline tests/test_asn1.ml]: I do indeed, I'll fix that right away.

File: tests/test_asn1.ml [hunk 3]
```diff
@@ -273,8 +273,80 @@ let ec_suite =
...
+  let cmp = Z.equal in
+  let printer = Z.to_string in
+  let test_params expected real ctxt =
+    let open Params in
+    assert_equal ~ctxt ~cmp ~printer ~msg:"p" expected.p real.p;
+    assert_equal ~ctxt ~cmp ~printer ~msg:"g" expected.g real.g;
+    assert_equal
+      ~ctxt
+      ~cmp:
+        (fun a b ->
```
Comment (NathanReb) [inline tests/test_asn1.ml]: Could you extract this function in `Test_util` ?

Comment (jcm-cryptosense) [inline tests/test_asn1.ml]: The function to compare two `option`s?

**Comment by NathanReb** [comment]
I'd say edit `CHANGELOG.md` and we're good !

**Comment by NathanReb** [comment]
Thank you ! Merging !

**Pull request merged**